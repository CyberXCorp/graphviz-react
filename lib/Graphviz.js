import { graphviz } from 'd3-graphviz';
import * as React from 'react';
export class Graphviz extends React.Component {
    constructor(props) {
        super(props);
        this.render = () => {
            return (React.createElement("div", { id: this.id }));
        };
        this.componentDidMount = () => {
            this.appendWasmScript();
            this.renderGraph();
        };
        this.componentDidUpdate = () => {
            this.renderGraph();
        };
        this.renderGraph = () => {
            graphviz('#' + this.id)
                .options(this.options())
                .renderDot(this.props.dot);
        };
        this.appendWasmScript = () => {
            const script = document.createElement("script");
            script.src = "https://unpkg.com/@hpcc-js/wasm/dist/index.min.js";
            script.type = "application/javascript/";
            document.body.appendChild(script);
        };
        this.options = () => {
            if (!this.props.options) {
                return Graphviz.defaultOptions;
            }
            const options = Graphviz.defaultOptions;
            for (const option of Object.keys(this.props.options)) {
                options[option] = this.props.options[option];
            }
            return options;
        };
        this.id = "graphviz" + Graphviz.count;
        Graphviz.count++;
    }
}
Graphviz.count = 0;
Graphviz.defaultOptions = {
    fit: true,
    height: 500,
    width: 500,
    zoom: false,
};
//# sourceMappingURL=Graphviz.js.map