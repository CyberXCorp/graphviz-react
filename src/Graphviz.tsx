import * as React from 'react';
import { graphviz, GraphvizOptions } from 'd3-graphviz';

export class Graphviz extends React.Component<IGraphvizProps, any> {
  private static count = 0;

  private static defaultOptions: GraphvizOptions = {
    fit: true,
    height: 500,
    width: 500,
    zoom: false,
  };

  private id: string;

  constructor(props: IGraphvizProps) {
    super(props);
    this.id = `graphviz${Graphviz.count}`;
    Graphviz.count += 1;
  }

  public render = (): JSX.Element => (
    <div id={this.id} />
  );

  public componentDidMount = () => {
    this.appendWasmScript();
    this.renderGraph();
  };

  public componentDidUpdate = () => {
    this.renderGraph();
  };

  private renderGraph = () => {
    const { dot, options } = this.props;
    graphviz(`#${this.id}`)
      .options({
        ...Graphviz.defaultOptions,
        ...options || {},
      })
      .renderDot(dot);
  };

  private appendWasmScript = () => {
    const script = document.createElement('script');

    script.src = 'https://unpkg.com/@hpcc-js/wasm/dist/index.min.js';
    script.type = 'application/javascript/';

    document.body.appendChild(script);
  };
}

export interface IGraphvizProps extends React.ClassAttributes<Graphviz> {
  dot: string;
  options?: GraphvizOptions;
}
