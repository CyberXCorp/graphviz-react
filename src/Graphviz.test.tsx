import { render } from '@testing-library/react';
import * as React from 'react';
import { Graphviz } from './Graphviz';

/**
 * Tests a given DOT string.
 * @param dot the DOT string to test.
 */
const test = (dot: string) => {
  describe(dot, () => {
    it('matches expected baseElement snapshot', () => {
      const { baseElement } = render(<Graphviz dot={dot} />);
      expect(baseElement).toMatchSnapshot();
    });

    it('matches expected container snapshot', () => {
      const { container } = render(<Graphviz dot={dot} />);
      expect(container).toMatchSnapshot();
    });
  });
};

describe('<Graphviz />', () => {
  beforeAll(() => {
    // Mock out the `getTotalLength` method for SVG as this is not currently implemented by JSDOM.
    // See: https://github.com/jsdom/jsdom/issues/1330 & https://github.com/jsdom/jsdom/issues/1423
    // @ts-ignore
    SVGElement.prototype.getTotalLength = jest.fn();
  });

  test('graph { a -- b }');
  test('digraph { a -> b }');
});
