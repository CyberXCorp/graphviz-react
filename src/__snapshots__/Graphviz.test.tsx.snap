// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Graphviz /> digraph { a -> b } matches expected baseElement snapshot 1`] = `
<body>
  <div>
    <div
      id="graphviz2"
    >
      <svg
        height="500"
        viewBox="0.00 0.00 62.00 116.00"
        width="500"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        

        <g
          class="graph"
          id="graph0"
          transform="scale(1 1) rotate(0) translate(4 112)"
        >
          

          <title>
            %0
          </title>
          

          <polygon
            fill="#ffffff"
            points="-4,4 -4,-112 58,-112 58,4 -4,4"
            stroke="transparent"
          />
          

          <!-- a -->
          

          <g
            class="node"
            id="node1"
          >
            

            <title>
              a
            </title>
            

            <ellipse
              cx="27"
              cy="-90"
              fill="none"
              rx="27"
              ry="18"
              stroke="#000000"
            />
            

            <text
              fill="#000000"
              font-family="Times,serif"
              font-size="14.00"
              text-anchor="middle"
              x="27"
              y="-85.8"
            >
              a
            </text>
            

          </g>
          

          <!-- b -->
          

          <g
            class="node"
            id="node2"
          >
            

            <title>
              b
            </title>
            

            <ellipse
              cx="27"
              cy="-18"
              fill="none"
              rx="27"
              ry="18"
              stroke="#000000"
            />
            

            <text
              fill="#000000"
              font-family="Times,serif"
              font-size="14.00"
              text-anchor="middle"
              x="27"
              y="-13.8"
            >
              b
            </text>
            

          </g>
          

          <!-- a&#45;&gt;b -->
          

          <g
            class="edge"
            id="edge1"
          >
            

            <title>
              a-&gt;b
            </title>
            

            <path
              d="M27,-71.8314C27,-64.131 27,-54.9743 27,-46.4166"
              fill="none"
              stroke="#000000"
            />
            

            <polygon
              fill="#000000"
              points="30.5001,-46.4132 27,-36.4133 23.5001,-46.4133 30.5001,-46.4132"
              stroke="#000000"
            />
            

          </g>
          

        </g>
        

      </svg>
    </div>
  </div>
</body>
`;

exports[`<Graphviz /> digraph { a -> b } matches expected container snapshot 1`] = `
<div>
  <div
    id="graphviz3"
  >
    <svg
      height="500"
      viewBox="0.00 0.00 62.00 116.00"
      width="500"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      

      <g
        class="graph"
        id="graph0"
        transform="scale(1 1) rotate(0) translate(4 112)"
      >
        

        <title>
          %0
        </title>
        

        <polygon
          fill="#ffffff"
          points="-4,4 -4,-112 58,-112 58,4 -4,4"
          stroke="transparent"
        />
        

        <!-- a -->
        

        <g
          class="node"
          id="node1"
        >
          

          <title>
            a
          </title>
          

          <ellipse
            cx="27"
            cy="-90"
            fill="none"
            rx="27"
            ry="18"
            stroke="#000000"
          />
          

          <text
            fill="#000000"
            font-family="Times,serif"
            font-size="14.00"
            text-anchor="middle"
            x="27"
            y="-85.8"
          >
            a
          </text>
          

        </g>
        

        <!-- b -->
        

        <g
          class="node"
          id="node2"
        >
          

          <title>
            b
          </title>
          

          <ellipse
            cx="27"
            cy="-18"
            fill="none"
            rx="27"
            ry="18"
            stroke="#000000"
          />
          

          <text
            fill="#000000"
            font-family="Times,serif"
            font-size="14.00"
            text-anchor="middle"
            x="27"
            y="-13.8"
          >
            b
          </text>
          

        </g>
        

        <!-- a&#45;&gt;b -->
        

        <g
          class="edge"
          id="edge1"
        >
          

          <title>
            a-&gt;b
          </title>
          

          <path
            d="M27,-71.8314C27,-64.131 27,-54.9743 27,-46.4166"
            fill="none"
            stroke="#000000"
          />
          

          <polygon
            fill="#000000"
            points="30.5001,-46.4132 27,-36.4133 23.5001,-46.4133 30.5001,-46.4132"
            stroke="#000000"
          />
          

        </g>
        

      </g>
      

    </svg>
  </div>
</div>
`;

exports[`<Graphviz /> graph { a -- b } matches expected baseElement snapshot 1`] = `
<body>
  <div>
    <div
      id="graphviz0"
    >
      <svg
        height="500"
        viewBox="0.00 0.00 62.00 116.00"
        width="500"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        

        <g
          class="graph"
          id="graph0"
          transform="scale(1 1) rotate(0) translate(4 112)"
        >
          

          <title>
            %0
          </title>
          

          <polygon
            fill="#ffffff"
            points="-4,4 -4,-112 58,-112 58,4 -4,4"
            stroke="transparent"
          />
          

          <!-- a -->
          

          <g
            class="node"
            id="node1"
          >
            

            <title>
              a
            </title>
            

            <ellipse
              cx="27"
              cy="-90"
              fill="none"
              rx="27"
              ry="18"
              stroke="#000000"
            />
            

            <text
              fill="#000000"
              font-family="Times,serif"
              font-size="14.00"
              text-anchor="middle"
              x="27"
              y="-85.8"
            >
              a
            </text>
            

          </g>
          

          <!-- b -->
          

          <g
            class="node"
            id="node2"
          >
            

            <title>
              b
            </title>
            

            <ellipse
              cx="27"
              cy="-18"
              fill="none"
              rx="27"
              ry="18"
              stroke="#000000"
            />
            

            <text
              fill="#000000"
              font-family="Times,serif"
              font-size="14.00"
              text-anchor="middle"
              x="27"
              y="-13.8"
            >
              b
            </text>
            

          </g>
          

          <!-- a&#45;&#45;b -->
          

          <g
            class="edge"
            id="edge1"
          >
            

            <title>
              a--b
            </title>
            

            <path
              d="M27,-71.8314C27,-61 27,-47.2876 27,-36.4133"
              fill="none"
              stroke="#000000"
            />
            

          </g>
          

        </g>
        

      </svg>
    </div>
  </div>
</body>
`;

exports[`<Graphviz /> graph { a -- b } matches expected container snapshot 1`] = `
<div>
  <div
    id="graphviz1"
  >
    <svg
      height="500"
      viewBox="0.00 0.00 62.00 116.00"
      width="500"
      xmlns="http://www.w3.org/2000/svg"
      xmlns:xlink="http://www.w3.org/1999/xlink"
    >
      

      <g
        class="graph"
        id="graph0"
        transform="scale(1 1) rotate(0) translate(4 112)"
      >
        

        <title>
          %0
        </title>
        

        <polygon
          fill="#ffffff"
          points="-4,4 -4,-112 58,-112 58,4 -4,4"
          stroke="transparent"
        />
        

        <!-- a -->
        

        <g
          class="node"
          id="node1"
        >
          

          <title>
            a
          </title>
          

          <ellipse
            cx="27"
            cy="-90"
            fill="none"
            rx="27"
            ry="18"
            stroke="#000000"
          />
          

          <text
            fill="#000000"
            font-family="Times,serif"
            font-size="14.00"
            text-anchor="middle"
            x="27"
            y="-85.8"
          >
            a
          </text>
          

        </g>
        

        <!-- b -->
        

        <g
          class="node"
          id="node2"
        >
          

          <title>
            b
          </title>
          

          <ellipse
            cx="27"
            cy="-18"
            fill="none"
            rx="27"
            ry="18"
            stroke="#000000"
          />
          

          <text
            fill="#000000"
            font-family="Times,serif"
            font-size="14.00"
            text-anchor="middle"
            x="27"
            y="-13.8"
          >
            b
          </text>
          

        </g>
        

        <!-- a&#45;&#45;b -->
        

        <g
          class="edge"
          id="edge1"
        >
          

          <title>
            a--b
          </title>
          

          <path
            d="M27,-71.8314C27,-61 27,-47.2876 27,-36.4133"
            fill="none"
            stroke="#000000"
          />
          

        </g>
        

      </g>
      

    </svg>
  </div>
</div>
`;
